stock_orderpoint_generator/models/product.py

```diff
@@ -0,0 +1,123 @@
+# Copyright 2017 Camptocamp SA
+# Copyright 2019 David Vidal - Tecnativa
+# Copyright 2020 Víctor Martínez - Tecnativa
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html)
+
+from collections import OrderedDict
+
+from odoo import api, fields, models
+
+
+class ProductProduct(models.Model):
+    _inherit = "product.product"
+
+    auto_orderpoint_template_ids = fields.Many2many(
+        comodel_name="stock.warehouse.orderpoint.template",
+        string="Automatic Reordering Rules",
+        domain=[("auto_generate", "=", True)],
+        help="When one or several automatic reordering rule is selected, "
+        "a Scheduled Action will automatically generate or update "
+        "the reordering rules of the product.",
+    )
+
+    @api.model
+    def create(self, vals):
+        record = super().create(vals)
+        if vals.get("auto_orderpoint_template_ids"):
+            record.auto_orderpoint_template_ids.create_orderpoints(record)
+        return record
+
+    def write(self, vals):
+        result = super().write(vals)
+        if vals.get("auto_orderpoint_template_ids"):
+            orderpoint_templates = self.mapped("auto_orderpoint_template_ids")
+            orderpoint_templates.create_orderpoints(self)
+        return result
+
+    def _compute_historic_quantities_dict(
+        self, location_id=False, from_date=False, to_date=False
+    ):
+        """Returns a dict of products with a dict of historic moves as for
+           a list of historic stock values resulting from those moves. If
+           a location_id is passed, we can restrict it to such location"""
+        location = location_id and location_id.id
+        domain_quant_loc, domain_move_in_loc, domain_move_out_loc = self.with_context(
+            location=location
+        )._get_domain_locations()
+        if not to_date:
+            to_date = fields.Datetime.now()
+        domain_move_in = domain_move_out = [
```

Comment:
IMHO this function is too long. Please separate some logical parts (domain creation, product_move_dict creation).

Reply:
Ok, thanks, separated it now in different functions.